File: go/cmd/command.go
package main

import (
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"strings"

	"ericc-ch/shittycannon/cannon"
	"ericc-ch/shittycannon/report"
	"ericc-ch/shittycannon/runoptions"

	"github.com/spf13/cobra"
)

const (
	defaultConnections = 10
	defaultDuration    = 10
	defaultTimeout     = 10
	defaultMethod      = "GET"
)

func checkPositive(name string, value int) error {
	if value <= 0 {
		return fmt.Errorf("%s: must be greater than 0, got %d", name, value)
	}
	return nil
}

func parseMethod(input string) (runoptions.Method, error) {
	upper := strings.ToUpper(input)
	switch runoptions.Method(upper) {
	case runoptions.MethodGet, runoptions.MethodPost, runoptions.MethodPut, runoptions.MethodDelete, runoptions.MethodPatch, runoptions.MethodHead, runoptions.MethodOptions, runoptions.MethodTrace:
		return runoptions.Method(upper), nil
	default:
		return "", fmt.Errorf("method: unsupported HTTP method %q", input)
	}
}

func parseTargetURL(input string) (*url.URL, error) {
	parsed, err := url.Parse(input)
	if err != nil {
		return nil, fmt.Errorf("url: %v", err)
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return nil, fmt.Errorf("url: must use http or https")
	}
	if parsed.Host == "" {
		return nil, fmt.Errorf("url: must include a host")
	}
	return parsed, nil
}

func parseHeaders(pairs []string) (map[string]string, error) {
	headers := make(map[string]string, len(pairs))
	for _, pair := range pairs {
		key, value, ok := strings.Cut(pair, "=")
		if !ok || key == "" {
			return nil, fmt.Errorf("headers: %q: expected key=value", pair)
		}
		headers[key] = value
	}
	return headers, nil
}

func newCommand() *cobra.Command {
	var (
		connections int
		duration    int
		amount      int
		method      string
		timeout     int
		body        string
		input       string
		headerPairs []string
		asJSON      bool
		latency     bool
	)

	cmd := &cobra.Command{
		Use:   "shittycannon [url]",
		Short: "HTTP/1 load tester (autocannon subset)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			useAmount := cmd.Flags().Changed("amount")
			if err := checkPositive("connections", connections); err != nil {
				return err
			}
			if err := checkPositive("timeout", timeout); err != nil {
				return err
			}
			if useAmount {
				if err := checkPositive("amount", amount); err != nil {
					return err
				}
			} else if err := checkPositive("duration", duration); err != nil {
				return err
			}
			parsedURL, err := parseTargetURL(args[0])
			if err != nil {
				return err
			}
			parsedMethod, err := parseMethod(method)
			if err != nil {
				return err
			}
			headers, err := parseHeaders(headerPairs)
			if err != nil {
				return err
			}
			bodySet := cmd.Flags().Changed("body")
			inputSet := cmd.Flags().Changed("input")
			if bodySet && inputSet {
				return fmt.Errorf("use either -b/--body or -i/--input, not both")
			}
			var requestBody string
			if bodySet {
				requestBody = body
			}
			if inputSet {
				raw, readErr := os.ReadFile(input)
				if readErr != nil {
					return readErr
				}
				requestBody = string(raw)
			}
			var stop runoptions.Stop = runoptions.Duration{Seconds: duration}
			if useAmount {
				stop = runoptions.Amount{Requests: amount}
			}
			result := cannon.Run(runoptions.RunOptions{
				URL:            parsedURL,
				Connections:    connections,
				Stop:           stop,
				Method:         parsedMethod,
				Headers:        headers,
				Body:           requestBody,
				TimeoutSeconds: timeout,
			})
			if asJSON {
				encoded, encErr := json.Marshal(result)
				if encErr != nil {
					return encErr
				}
				_, _ = fmt.Fprintln(cmd.OutOrStdout(), string(encoded))
				return nil
			}
			_, _ = fmt.Fprintln(cmd.OutOrStdout(), report.FormatText(result, latency))
			return nil
		},
	}

	cmd.Flags().IntVarP(&connections, "connections", "c", defaultConnections, "The number of concurrent connections to use. default: 10.")
	cmd.Flags().IntVarP(&duration, "duration", "d", defaultDuration, "The number of seconds to run the autocannon. default: 10.")
	cmd.Flags().IntVarP(&amount, "amount", "a", 0, "The number of requests to make before exiting the benchmark. If set, duration is ignored.")
	cmd.Flags().StringVarP(&method, "method", "m", defaultMethod, "The HTTP method to use. default: 'GET'.")
	cmd.Flags().IntVarP(&timeout, "timeout", "t", defaultTimeout, "The number of seconds before timing out and resetting a connection. default: 10")
	cmd.Flags().StringVarP(&body, "body", "b", "", "The body of the request.")
	cmd.Flags().StringVarP(&input, "input", "i", "", "The body of the request. See '-b/body' for more details.")
	cmd.Flags().StringArrayVarP(&headerPairs, "headers", "H", nil, "The request headers.")
	cmd.Flags().BoolVarP(&asJSON, "json", "j", false, "Print the output as newline delimited JSON. This will cause the progress bar and results not to be rendered. default: false.")
	cmd.Flags().BoolVarP(&latency, "latency", "l", false, "Print all the latency data. default: false.")
	return cmd
}

File: typescript-effect/src/command.ts
import { Console, Effect, FileSystem, Option, Schema } from "effect"
import { Argument, CliError, Command, Flag } from "effect/unstable/cli"
import { runCannon } from "./cannon.ts"
import { formatText } from "./report.ts"
import { Body, type RunOptions, Stop } from "./run-options.ts"

const Disabled = false
const DefaultConnections = 10
const DefaultDuration = 10
const DefaultTimeout = 10
const DefaultMethod = "GET"

const PositiveInt = Schema.Int.check(Schema.isGreaterThan(0))

export const shittycannon = Command.make(
  "shittycannon",
  {
    url: Argument.string("url").pipe(
      Argument.withDescription("HTTP or HTTPS URL to benchmark"),
      Argument.mapEffect((input) =>
        Schema.decodeUnknownEffect(
          Schema.URLFromString.check(
            Schema.makeFilter((url: URL) => {
              if (url.protocol === "http:" || url.protocol === "https:") {
                return true
              }
              return "URL must use http or https"
            }),
          ),
        )(input).pipe(
          Effect.mapError((error) =>
            new CliError.InvalidValue({
              option: "url",
              value: input,
              expected: error.message,
              kind: "argument",
            })
          ),
        )
      ),
    ),
    connections: Flag.integer("connections").pipe(
      Flag.withAlias("c"),
      Flag.withDescription(
        "The number of concurrent connections to use. default: 10.",
      ),
      Flag.withDefault(DefaultConnections),
      Flag.withSchema(PositiveInt),
    ),
    duration: Flag.integer("duration").pipe(
      Flag.withAlias("d"),
      Flag.withDescription(
        "The number of seconds to run the autocannon. default: 10.",
      ),
      Flag.withDefault(DefaultDuration),
      Flag.withSchema(PositiveInt),
    ),
    amount: Flag.integer("amount").pipe(
      Flag.withAlias("a"),
      Flag.withDescription(
        "The number of requests to make before exiting the benchmark. If set, duration is ignored.",
      ),
      Flag.withSchema(PositiveInt),
      Flag.optional,
    ),
    method: Flag.string("method").pipe(
      Flag.withAlias("m"),
      Flag.withDescription("The HTTP method to use. default: 'GET'."),
      Flag.withDefault(DefaultMethod),
      Flag.mapEffect((input) =>
        Schema.decodeUnknownEffect(
          Schema.Literals([
            "GET",
            "POST",
            "PUT",
            "DELETE",
            "PATCH",
            "HEAD",
            "OPTIONS",
            "TRACE",
          ]),
        )(input.toUpperCase()).pipe(
          Effect.mapError((error) =>
            new CliError.InvalidValue({
              option: "method",
              value: input,
              expected: error.message,
              kind: "flag",
            })
          ),
        )
      ),
    ),
    timeout: Flag.integer("timeout").pipe(
      Flag.withAlias("t"),
      Flag.withDescription(
        "The number of seconds before timing out and resetting a connection. default: 10",
      ),
      Flag.withDefault(DefaultTimeout),
      Flag.withSchema(PositiveInt),
    ),
    body: Flag.string("body").pipe(
      Flag.withAlias("b"),
      Flag.withDescription("The body of the request."),
      Flag.optional,
    ),
    input: Flag.file("input", { mustExist: true }).pipe(
      Flag.withAlias("i"),
      Flag.withDescription(
        "The body of the request. See '-b/body' for more details.",
      ),
      Flag.optional,
    ),
    headers: Flag.keyValuePair("headers").pipe(
      Flag.withAlias("H"),
      Flag.withDescription("The request headers."),
      Flag.withDefault({}),
    ),
    json: Flag.boolean("json").pipe(
      Flag.withAlias("j"),
      Flag.withDescription(
        "Print the output as newline delimited JSON. This will cause the progress bar and results not to be rendered. default: false.",
      ),
      Flag.withDefault(Disabled),
    ),
    latency: Flag.boolean("latency").pipe(
      Flag.withAlias("l"),
      Flag.withDescription("Print all the latency data. default: false."),
      Flag.withDefault(Disabled),
    ),
  },
  Effect.fn("shittycannon")(function* (config) {
    if (Option.isSome(Option.all([config.body, config.input]))) {
      return yield* new CliError.UserError({
        cause: "use either -b/--body or -i/--input, not both",
        userMessage: "use either -b/--body or -i/--input, not both",
      })
    }
    const runOptions: RunOptions = {
      url: config.url,
      connections: config.connections,
      stop: Option.match(config.amount, {
        onNone: () => Stop.Duration({ seconds: config.duration }),
        onSome: (requests) => Stop.Amount({ requests }),
      }),
      method: config.method,
      headers: config.headers,
      body: yield* Option.match(config.body, {
        onSome: (value) => Effect.succeed(Body.Text({ value })),
        onNone: () =>
          Option.match(config.input, {
            onSome: (path) =>
              Effect.gen(function* () {
                const fs = yield* FileSystem.FileSystem
                const value = yield* fs.readFileString(path)
                return Body.Text({ value })
              }),
            onNone: () => Effect.succeed(Body.Empty({})),
          }),
      }),
      timeoutSeconds: config.timeout,
    }
    const report = yield* runCannon(runOptions)
    if (config.json) {
      yield* Console.log(JSON.stringify(report))
      return
    }
    yield* Console.log(formatText(report, config.latency))
  }),
).pipe(
  Command.withDescription("HTTP/1 load tester (autocannon subset)"),
)
